### PR TITLE
Auth part with dsl drilling

### LINKED ISSUE #96: Make a YAML DSL to extract tokens into request header entries

### LINKED ISSUE #115: auth/authz - Make sure there is no reference to owner/tenant-uuid

### PR DESCRIPTION
**Pull request by Emulebest**
resolves #115 
closes #96

### CODE REVIEW

File: __tests__/entity_api_auth.test.ts [hunk 1]
```diff
@@ -5,7 +5,7 @@ const url = require("url");
 const queryString = require("query-string");
 import { Metadata, Spec } from "../src/core";
 import { Provider } from "../src/papiea";
-import { getProviderWithSpecOnlyEnitityKindWithOperations } from "./test_data_factory";
+import { getProviderWithSpecOnlyEnitityKindWithOperations, loadYaml } from "./test_data_factory";
```
Comment (adolgarev) [inline __tests__/entity_api_auth.test.ts]: add another function to test_data_factory which will return provider with operations and auth data in it

File: __tests__/auth_user_data.test.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        try {
+            const the_headers = evaluate_headers(env, headers);
+            expect(the_headers).toEqual(
+                {
+                    "tenant-email": "shlomi.vaknin@nutanix.com",
+                    "tenant-id": "262505c1-83e3-4158-a153-e3377afd979b",
+                    "tenant-fname": "shlomi",
+                    "tenant-lname": "vaknin",
+                    "tenant-role": "account-admin",
+                    "owner": "shlomi.vaknin@nutanix.com"
```
Comment (adolgarev) [inline __tests__/auth_user_data.test.ts]: believe this should be anonymized

File: __tests__/auth.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+          
+    # headers is the only entry that is mandatory in this section
+    # All lines listed will be added to the request's header. 
+    headers:
+      # These will use the same syntax as role above
+      tenant-email: ^id_token.email
+      tenant-id:    ^id_token.default_tenant
+      tenant-fname: ^id_token.given_name
+      tenant-lname: ^id_token.last_name
+      owner: ^id_token.sub
```
Comment (adolgarev) [inline __tests__/auth.yaml]: headers are passed to provider, do not confuse with user_info which is used in casbin model for authorization. Thus owner should be in user_inof.owner. Also user_info should have tenent field. Also I Io not think we need to store access_token and id_token in user_info, only owner, tenant and headers.

File: src/@types/node-json-color-stringify.d.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+/** Declaration file generated by dts-gen */
+
+export function colorStringify(obj: any, filter: any, indent: any, level: any): any;
```
Comment (adolgarev) [inline src/@types/node-json-color-stringify.d.ts]: why is this needed?

File: src/auth/authn.ts [hunk 5]
```diff
@@ -16,6 +16,15 @@ export interface UserAuthInfo {
     headers?: { [key: string]: string; };
 }
 
+export function mapHeadersToFields(user_auth_info: UserAuthInfo) {
+    if (user_auth_info.headers === undefined) {
+        throw new UnauthorizedError();
+    }
+    for (let header in user_auth_info.headers) {
+        user_auth_info[header] = user_auth_info.headers[header]
```
Comment (adolgarev) [inline src/auth/authn.ts]: that is completely wrong, see my comment above

File: src/auth/oauth2.ts [hunk 6]
```diff
@@ -1,57 +1,52 @@
...
-                content = JSON.parse(atob(content_base64));
-            }
-            return { header, content };
+    const simple_oauth_config = {
+        client: {
+            id: oauth.client_id,
+            secret: oauth.client_secret
+        },
+        auth: {
+            tokenHost: `http://${oauth.auth_host}`,
```
Comment (adolgarev) [inline src/auth/oauth2.ts]: most if not all use httpS, scheme should be a part of auth_host

File: src/auth/oauth2.ts [hunk 7]
```diff
@@ -1,57 +1,52 @@
...
+function getUserInfoFromToken(token: any, oauth_description: any): UserAuthInfo {
+    const userInfo: UserAuthInfo = {};
+
+    const headers = oauth_description.oauth.user_info.headers;
+    let env = _.omit(oauth_description.oauth.user_info, ['headers']);
+
+    env.token = token.token;
+
+    const extracted_headers = evaluate_headers(env, headers);
+    extracted_headers.authorization = `Bearer ${token.token.access_token}`;
```
Comment (adolgarev) [inline src/auth/oauth2.ts]: this should not be here, this must be a part of DSL

File: src/auth/oauth2.ts [hunk 8]
```diff
@@ -72,7 +67,8 @@ export function createOAuth2Router(redirect_uri: string, signature: Signature, p
 
     router.use('/provider/:prefix/:version/auth/login', asyncHandler(async (req, res) => {
         const provider: Provider = await providerDb.get_provider(req.params.prefix, req.params.version);
-        const oauth2 = simpleOauthModule.create(provider.oauth2);
+        const converted_oauth = convertToSimpleOauth2(provider.oauth2);
+        const oauth2 = simpleOauthModule.create(converted_oauth);
```
Comment (adolgarev) [inline src/auth/oauth2.ts]: there should be 1 function that returns oauth2 given provider

File: src/auth/oauth2.ts [hunk 9]
```diff
@@ -92,14 +88,15 @@ export function createOAuth2Router(redirect_uri: string, signature: Signature, p
...
         try {
             const result = await oauth2.authorizationCode.getToken({
                 code,
                 redirect_uri
             });
             const token = oauth2.accessToken.create(result);
-            const userInfo = getUserInfoFromToken(token);
+            const userInfo = getUserInfoFromToken(token, provider.oauth2);
```
Comment (adolgarev) [inline src/auth/oauth2.ts]: it should take provider as an input

File: src/auth/oauth2.ts [hunk 10]
```diff
@@ -124,6 +121,9 @@ export function createOAuth2Router(redirect_uri: string, signature: Signature, p
         let userInfo: UserAuthInfo;
         try {
             userInfo = await signature.verify(token);
+
+            // TODO: this doesn't look good, maybe rework this completely or introduce functional approach
+            mapHeadersToFields(userInfo);
```
Comment (adolgarev) [inline src/auth/oauth2.ts]: this is bad, userInfo should have owner, tenant, headers at this point, no need to mapHeadersToFields, see my comments above

File: src/auth/user_data_evaluator.ts [hunk 11]
```diff
@@ -0,0 +1,125 @@
+import * as yaml from 'js-yaml'
```
Comment (adolgarev) [inline src/auth/user_data_evaluator.ts]: cannot understand by the name of the file what is inside.

File: __tests__/auth.yaml (hunk 3 above)
Comment (adolgarev) [reply __tests__/auth.yaml]: I  believe you missed a point, the only thing that should have been done is replace getUserInfoFromToken(token) with getUserInfoFromToken(token, provider) which applies DSL stored in provider to replace hardcoded logic in there, it should return THE SAME userInfo as before

Comment (Emulebest) [reply __tests__/auth.yaml]: https://github.com/nutanix/papiea-js/issues/115

File: __tests__/auth_user_data.test.ts (hunk 2 above)
Comment (Emulebest) [reply __tests__/auth_user_data.test.ts]: Please elaborate on that, didn't get it

File: src/@types/node-json-color-stringify.d.ts (hunk 4 above)
Comment (Emulebest) [reply src/@types/node-json-color-stringify.d.ts]: This was used in the DSL part to highlight errors

File: __tests__/auth.yaml (hunk 3 above)
Comment (Emulebest) [reply __tests__/auth.yaml]: The point that we discussed with @shlomi-nx yesterday was about being agnostic and that owner and tenant are provider specific

File: src/auth/authn.ts (hunk 5 above)
Comment (Emulebest) [reply src/auth/authn.ts]: If user info and headers are separate entities why do we need to encode user_info as JWT with `headers?: any` as an optional param?

File: __tests__/auth.yaml (hunk 3 above)
Comment (adolgarev) [reply __tests__/auth.yaml]: @Emulebest 
yes, they are provider specific and should be mentioned only in this DSL and casbin model, this doesn't mean you should move everything under headers. The process is the following: DSL fill in UserAuthInfo which is passed to casbin model/policy and is opaque for the rest of papies. If you search for owner or tenant mentions in the code you'll find only one place - getUserInfoFromToken function

File: __tests__/test_data_factory.ts [hunk 12]
```diff
@@ -81,3 +81,33 @@ export function getProviderWithSpecOnlyEnitityKindWithOperations(procedure_callb
     provider.kinds[0].procedures[geolocationComputeProceduralSignature.name] = geolocationComputeProceduralSignature;
     return provider;
 }
+
+export function getProviderWithSpecOnlyEnitityKindWithOperationsOauthDescription(procedure_callback: Provider_Callback_URL): Provider {
```
Comment (adolgarev) [inline __tests__/test_data_factory.ts]: getProviderWithSpecOnlyEnitityKindWithOperationsAndOAuthDescription

File: src/auth/oauth2.ts [hunk 13]
```diff
@@ -35,16 +36,15 @@ function convertToSimpleOauth2(description: any) {
     return simple_oauth_config;
 }
 
-function getUserInfoFromToken(token: any, oauth_description: any): UserAuthInfo {
-    const userInfo: UserAuthInfo = {};
-
-    const headers = oauth_description.oauth.user_info.headers;
-    let env = _.omit(oauth_description.oauth.user_info, ['headers']);
+function getOauth2(provider: Provider) {
```
Comment (adolgarev) [inline src/auth/oauth2.ts]: OAuth

File: __tests__/test_data_factory.ts (hunk 12 above)
Comment (Emulebest) [reply __tests__/test_data_factory.ts]: I think we would need to integrate builder later not to get messy with the naming patterns

File: src/auth/authn.ts (hunk 5 above)
Comment (adolgarev) [reply src/auth/authn.ts]: because we need to restore this headers from token passed in to later pass them to procedure

File: __tests__/entity_api_auth.test.ts [hunk 14]
```diff
@@ -5,7 +5,11 @@ const url = require("url");
 const queryString = require("query-string");
 import { Metadata, Spec } from "../src/core";
 import { Provider } from "../src/papiea";
-import { getProviderWithSpecOnlyEnitityKindWithOperations } from "./test_data_factory";
+import {
+    getProviderWithSpecOnlyEnitityKindWithOperations,
```
Comment (adolgarev) [inline __tests__/entity_api_auth.test.ts]: not used

File: __tests__/entity_api_auth.test.ts [hunk 15]
```diff
@@ -5,7 +5,11 @@ const url = require("url");
...
+    getProviderWithSpecOnlyEntityKindWithOperationsAndOAuth2Description,
+    loadYaml
```
Comment (adolgarev) [inline __tests__/entity_api_auth.test.ts]: not used

File: src/auth/oauth2.ts [hunk 16]
```diff
@@ -1,57 +1,52 @@
...
+function getOAuth2(provider: Provider) {
+    const converted_oauth = convertToSimpleOauth2(provider.oauth2);
+    return simpleOauthModule.create(converted_oauth);
+}
+
+function getUserInfoFromToken(token: any, provider: Provider): UserAuthInfo {
+
+    const extracted_headers = extract_headers(token, provider.oauth2);
+    const extracted_user_props = extract_user_props(token, provider.oauth2);
+    console.log(extracted_user_props);
```
Comment (adolgarev) [inline src/auth/oauth2.ts]: remove console.log

File: src/auth/user_data_evaluator.ts [hunk 17]
```diff
@@ -0,0 +1,144 @@
+import * as _ from 'lodash'
+import * as JSON from 'node-json-color-stringify'
+import * as color from 'colors'
+
+const str = (x: any) => JSON.colorStringify(x, null, 2);
```
Comment (adolgarev) [inline src/auth/user_data_evaluator.ts]: please get rid of colors, this is good for running this from cli, but not on prod

File: src/auth/user_data_evaluator.ts [hunk 18]
```diff
@@ -0,0 +1,144 @@
+import * as _ from 'lodash'
+import * as JSON from 'node-json-color-stringify'
```
Comment (adolgarev) [inline src/auth/user_data_evaluator.ts]: think not needed really

File: src/auth/user_data_evaluator.ts [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+import * as color from 'colors'
```
Comment (adolgarev) [inline src/auth/user_data_evaluator.ts]: think not needed really

File: src/auth/user_data_evaluator.ts [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+export function extract_headers(token: any, oauth_description: any): any {
+    const headers = oauth_description.oauth.user_info.headers;
+    let env = _.omit(oauth_description.oauth.user_info, ['headers']);
+
+    env.token = token.token;
+
+    const extracted_headers = evaluate_headers(env, headers);
+    extracted_headers.authorization = `Bearer ${token.token.access_token}`;
```
Comment (adolgarev) [inline src/auth/user_data_evaluator.ts]: this must be a part of DSL

File: src/auth/user_data_evaluator.ts [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+    return _.mapValues(user_identifiers, (v: any) => deref(env, v))
+}
+
+export function extract_headers(token: any, oauth_description: any): any {
+    const headers = oauth_description.oauth.user_info.headers;
+    let env = _.omit(oauth_description.oauth.user_info, ['headers']);
+
+    env.token = token.token;
+
+    const extracted_headers = evaluate_headers(env, headers);
```
Comment (adolgarev) [inline src/auth/user_data_evaluator.ts]: this is one liner used only once, why to introduce a function?

File: src/@types/node-json-color-stringify.d.ts (hunk 4 above)
Comment (adolgarev) [reply src/@types/node-json-color-stringify.d.ts]: think not needed really

File: __tests__/auth_user_data.test.ts (hunk 2 above)
Comment (adolgarev) [reply __tests__/auth_user_data.test.ts]: remove real names from test data

File: __tests__/test_data_factory.ts (hunk 12 above)
Comment (adolgarev) [reply __tests__/test_data_factory.ts]: feel free to do so, not in this PR

File: src/auth/user_data_evaluator.ts (hunk 20 above)
Comment (shlomi-nx) [reply src/auth/user_data_evaluator.ts]: Yes, I see three options for this:

1. We could add a `$concat()` function to the dsl
    ```yaml
     headers:
       tenant-email: ^id_token.email
       authorization: $concat('Bearer ', ^id_token.access_token)
    ```
    - Even though concat may be useful in general, I think that for this use case it is much better to have something more specific

1. We could add a `$bearer()` function to the dsl
    ```yaml
     headers:
       tenant-email: ^id_token.email
       authorization: $bearer(^id_token.access_token)
    ```
    - This is probably the way to go, but it may be harder to implement. I don't think the dsl currently knows how to start with a function nor accept a reference as parameter

2. We could have `authorization` be a special key that automatically includes `Bearer`:
    ```yaml
     headers:
       tenant-email: ^id_token.email          // Regular key
       authorization: ^id_token.access_token  // Special authorization key
    ```
    - This is a tradeoff as its much easier to implement, but introduces exceptional cases to the DSL

File: src/auth/user_data_evaluator.ts (hunk 21 above)
Comment (shlomi-nx) [reply src/auth/user_data_evaluator.ts]: I originally left the one liner there, but then I thought that it may be easier for the uninitiated to understand at a glance what the one liner actually does if it has a name. So this is essentially a form of a comment. 

We can inline it but then maybe lets at least have 'evaluate_headers' as a comment.

File: src/auth/user_data_evaluator.ts [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+    return _.mapValues(headers, (v: any) => deref(env, v))
+}
+
+export function extract_user_props(token: any, oauth_description: any): any {
+    const user_identifiers = oauth_description.oauth.user_info.user_props;
+    let env = _.omit(oauth_description.oauth.user_info, ['user_props']);
+
+    env.token = token.token;
+
+    return _.mapValues(user_identifiers, (v: any) => deref(env, v))
```
Comment (shlomi-nx) [inline src/auth/user_data_evaluator.ts]: By the way, `evaluate_headers` could also be used here, which would make it be used twice ;)

File: src/auth/user_data_evaluator.ts [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+
+    // each part operates on the value returned by processing the previous part
+    return _.reduce(parts, part_processor, env)
+}
+
+export function evaluate_headers(env: Object, headers: Object) {
+    return _.mapValues(headers, (v: any) => deref(env, v))
+}
+
+export function extract_user_props(token: any, oauth_description: any): any {
```
Comment (shlomi-nx) [inline src/auth/user_data_evaluator.ts]: Also, it seems that `extract_user_props` and `extract_headers` are very similar. Perhaps they could be generalized and parametrized by the `headers` parameter?

File: __tests__/auth_user_data.test.ts (hunk 2 above)
Comment (shlomi-nx) [reply __tests__/auth_user_data.test.ts]: probably a good idea!

File: package.json [hunk 24]
```diff
@@ -19,17 +19,20 @@
   "dependencies": {
     "@types/express": "^4.16.0",
     "@types/js-yaml": "^3.11.4",
+    "@types/lodash": "^4.14.123",
     "@types/mongodb": "^3.1.17",
     "@types/node": "^10.12.9",
     "@types/uuid": "^3.4.4",
     "axios": "^0.18.0",
     "casbin": "^2.0.1",
+    "colors": "^1.3.3",
```
Comment (adolgarev) [inline package.json]: this is not needed

File: src/auth/oauth2.ts [hunk 25]
```diff
@@ -1,57 +1,51 @@
...
+    const converted_oauth = convertToSimpleOauth2(provider.oauth2);
+    return simpleOauthModule.create(converted_oauth);
+}
+
+function getUserInfoFromToken(token: any, provider: Provider): UserAuthInfo {
+
+    const extracted_headers = extract_property(token, provider.oauth2, "headers");
+    const extracted_user_props = extract_property(token, provider.oauth2, "user_props");
+
+    const userInfo: UserAuthInfo = {headers: extracted_headers, ...extracted_user_props};
```
Comment (shlomi-nx) [inline src/auth/oauth2.ts]: Why do we need this special `user_props`? Whats wrong with simply using `headers`?

File: __tests__/auth.yaml [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+  client_secret: YYY
+  auth_host: http://127.0.0.1:9002
+  authorize_uri: /oauth2/authorize    # relative to auth_host
+  revoke_uri: /oauth2/revoke          # relative to auth_host
+  token_uri: /oauth2/token            # relative to auth_host
+  redirect_uri: /procedure/redirect   # a provider static procedure
+  user_info:
+    access_token: ^token.access_token.$JWT.content  # how to get the access_token from the token
+    id_token: ^token.id_token.$JWT.content # how to get the access_token from the token
+    user_props:
```
Comment (adolgarev) [inline __tests__/auth.yaml]: I think this user_props is obsolete

File: __tests__/auth_user_data.test.ts (hunk 2 above)
Comment (adolgarev) [reply __tests__/auth_user_data.test.ts]: @Emulebest please remove real names

File: __tests__/auth_user_data.test.ts [hunk 27]
```diff
@@ -0,0 +1,192 @@
...
+    env.token = token.token;
+
+    test("Headers", (done) => {
+        try {
+
+            //evaluate headers
+            const the_headers = _.mapValues(headers, (v: any) => deref(env, v));
+            expect(the_headers).toEqual(
+                {
+                    "tenant-email": "shlomi.vaknin@nutanix.com",
```
Comment (adolgarev) [inline __tests__/auth_user_data.test.ts]: Please remove real names

File: __tests__/casbin_authorizer.test.ts [hunk 28]
```diff
@@ -6,7 +6,7 @@ import { resolve } from "path";
 
 
 describe("Casbin authorizer tests", () => {
-    const pathToModel: string = resolve(__dirname, "../src/auth/default_provider_model.txt");
+    const pathToModel: string = resolve(__dirname, "./test_provider_model.txt");
```
Comment (adolgarev) [inline __tests__/casbin_authorizer.test.ts]: we should test against default model, this test is to show how default model behaves

File: src/auth/default_provider_model.txt [hunk 29]
```diff
@@ -12,5 +12,5 @@ g2 = _, _
 e = some(where (p.eft == allow)) && !some(where (p.eft == deny))
 
 [matchers]
-m = (r.sub.owner == p.sub || g(r.sub.owner, p.sub) || g2(p.sub, "default")) && ((p.owner == "owner" && r.sub.owner == r.obj.metadata.extension.owner) || (p.owner == "tenant" && r.sub.tenant == r.obj.metadata.extension.tenant_uuid) || p.owner == "anyone" ) && (p.obj == r.obj.metadata.kind || p.obj == "*") && (r.act == p.act || p.act == "*")
+m = (r.sub.headers.owner == p.sub || g(r.sub.headers.owner, p.sub) || g2(p.sub, "default")) && ((p.owner == "owner" && r.sub.headers.owner == r.obj.metadata.extension.owner) || (p.owner == "tenant" && r.sub.headers.tenant-id == r.obj.metadata.extension.tenant_uuid) || p.owner == "anyone" ) && (p.obj == r.obj.metadata.kind || p.obj == "*") && (r.act == p.act || p.act == "*")
```
Comment (adolgarev) [inline src/auth/default_provider_model.txt]: Do not think headers should be referenced here, I thought that headers should become sub, that is UserAuthInfo will have properties tenant-id and owner

File: src/auth/oauth2.ts [hunk 30]
```diff
@@ -1,57 +1,51 @@
...
+
+function getOAuth2(provider: Provider) {
+    const converted_oauth = convertToSimpleOauth2(provider.oauth2);
+    return simpleOauthModule.create(converted_oauth);
+}
+
+function getUserInfoFromToken(token: any, provider: Provider): UserAuthInfo {
+
+    const extracted_headers = extract_property(token, provider.oauth2, "headers");
+    const extracted_user_props = extract_property(token, provider.oauth2, "user_props");
```
Comment (adolgarev) [inline src/auth/oauth2.ts]: I thought that headers should become sub, that is UserAuthInfo will have properties tenant-id, owner, etc.
